From 20ff7064ed4c9a7a0f014ed099151ae8887ff65f Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Thu, 16 May 2024 12:11:36 +0200
Subject: [PATCH 0881/1695] plat-stm32mp2: scmi_server: add SCMI CortexM33
 agent and channel

Add a new SCMI agent dedicated to M33 with only one channel.
We will share, with this agent, the same clocks and resets
than the A35 agent.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d5f2a006e1dbafb164d667dcb9b8cc735a828344 plat-stm32mp2: scmi_server: add SCMI CortexM33 agent and channel]
Co-developed-by: Raphael Gallais-Pou <raphael.gallais-pou@foss.st.com>
Signed-off-by: Raphael Gallais-Pou <raphael.gallais-pou@foss.st.com>
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I700cc65cee392c8a876f6500952936b13009e2e4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/380385
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     | 42 ++++++++++++++++---
 1 file changed, 37 insertions(+), 5 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index 975bb9928..4388a3fde 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -469,7 +469,7 @@ static struct stm32_scmi_pd stm32_scmi_pd[] = {
 };
 #endif
 
-static const struct channel_resources scmi_channel[] = {
+static const struct channel_resources scmi_channel_a35_ns[] = {
 	[0] = {
 		.clock = stm32_scmi_clock,
 		.clock_count = ARRAY_SIZE(stm32_scmi_clock),
@@ -482,6 +482,15 @@ static const struct channel_resources scmi_channel[] = {
 	},
 };
 
+static const struct channel_resources scmi_channel_m33_ns[] = {
+	[0] = {
+		.clock = stm32_scmi_clock,
+		.clock_count = ARRAY_SIZE(stm32_scmi_clock),
+		.rd = stm32_scmi_reset,
+		.rd_count = ARRAY_SIZE(stm32_scmi_reset),
+	},
+};
+
 /*
  * Platform clocks exposed with SCMI
  */
@@ -697,17 +706,34 @@ static TEE_Result scmi_scpfw_cfg_early_init(void)
 	unsigned int j = 0;
 
 	scpfw_cfg.agent_count = 1;
+
+	if (IS_ENABLED(CFG_SCMI_CORTEXM_AGENT))
+		scpfw_cfg.agent_count += 1;
+
 	scpfw_cfg.agent_config = calloc(scpfw_cfg.agent_count,
 					sizeof(*scpfw_cfg.agent_config));
 
-	scpfw_cfg.agent_config[index].name = "agent";
+	scpfw_cfg.agent_config[index].name = "agent_a35";
 	scpfw_cfg.agent_config[index].agent_id = index + 1;
-	scpfw_cfg.agent_config[index].channel_count = ARRAY_SIZE(scmi_channel);
+	scpfw_cfg.agent_config[index].channel_count =
+		ARRAY_SIZE(scmi_channel_a35_ns);
 	scpfw_cfg.agent_config[index].channel_config =
 		calloc(scpfw_cfg.agent_config[index].channel_count,
 		       sizeof(*scpfw_cfg.agent_config[index].channel_config));
 	index++;
 
+	if (IS_ENABLED(CFG_SCMI_CORTEXM_AGENT)) {
+		scpfw_cfg.agent_config[index].name = "agent_m33";
+		scpfw_cfg.agent_config[index].agent_id = index + 1;
+		scpfw_cfg.agent_config[index].channel_count =
+			ARRAY_SIZE(scmi_channel_m33_ns);
+		scpfw_cfg.agent_config[index].channel_config =
+			calloc(scpfw_cfg.agent_config[index].channel_count,
+			       sizeof(*scpfw_cfg.agent_config[index]
+						.channel_config));
+		index++;
+	}
+
 	assert(scpfw_cfg.agent_count == index);
 
 	for (i = 0; i < scpfw_cfg.agent_count; i++)
@@ -803,9 +829,15 @@ static TEE_Result scmi_scpfw_cfg_init(void)
 	get_scmi_clocks();
 	get_scmi_resets();
 
-	scmi_scpfw_cfg_init_channel("channel",
+	scmi_scpfw_cfg_init_channel("channel_a35_ns",
 				    &scpfw_cfg.agent_config[0],
-				    scmi_channel);
+				    scmi_channel_a35_ns);
+
+	if (IS_ENABLED(CFG_SCMI_CORTEXM_AGENT)) {
+		scmi_scpfw_cfg_init_channel("channel_m33_ns",
+					    &scpfw_cfg.agent_config[1],
+					    scmi_channel_m33_ns);
+	}
 
 	/* DVFS will be populated from cpu_opp driver */
 
-- 
2.47.2

